### PR TITLE
Fix TileLayer lag during fling/animated movement

### LINKED ISSUE #1245: [BUG] TileLayer lag/jitter
**Describe The Bug**
When doing animated panning in flutter web using the example app (Animated Mapcontroller page) the markers (and any other non tile layer) seem to jitter against the map tiles.

**Expected Behavior**
The markers and other non tile layers should move together with the tile layer.

**Screenshots & Recordings**

https://user-images.githubusercontent.com/3683599/169871653-331472de-b182-4228-8923-ed9ce139590c.mp4

**Additional Information**
In the example I slowed the animation to 5 seconds to make the issue more obvious. I have **not** observed this issue when running the example on the Android emulator, only with flutter web on chrome. This is encountered both using the latest published version of flutter_map as well as `master`. Flutter version `2.10.5`.

When I run in release mode the effect is more subtle but there is still jittering. In any case the movement of the markers and the map tiles should be synchronised so that any jitter/jank should affect them together.

---

**Doctors Report**

```shell
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.10.5, on macOS 11.5.2 20G95 darwin-x64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.65.0)
[✓] Connected device (2 available)
[✓] HTTP Host Availability
```

**To Reproduce**

1. Edit the example app to slow the animation of the Animated Mapcontroller page movement to 5 seconds. 
2. Run the example app with flutter web and go to the animated mapcontroller page.
3. Click the different location buttons and observe the jitter during panning.

---

**Severity**
This will help us to label the issue quicker and decide what needs attention first. Only choose fatal if the entire app crashes, otherwise choose non-fatal.

- [x] Non-Fatal
- [ ] Fatal / App Crashes

**Frequency/Rarity**
This will help us to label the issue quicker and decide what needs attention first.

- [ ] Once
- [ ] Uncommon
- [ ] Common
- [x] Always

**Applicable Platforms**
Only select those that you've tested on - one or more. If possible, test on a variety of platforms.

- [ ] Android
- [ ] iOS
- [x] Web
- [ ] Windows
- [ ] Others (beta platforms)

### PR DESCRIPTION
**Pull request by rorystephenson**
Fixes #1245 

When flinging or animating the moment of the map the TileLayer would lag behind which resulted in either jerkiness or other layers moving out of sync with the TileLayer depending on how fast the device is.

This PR should address that issue by moving the Tile translation calculation in to the builder rather than in a stream listener which called setState(). This means that translation is calculated during the build whilst updating/loading of tiles can still happen in the stream listener to prevent them from slowing down the build too much.

As part of these changes I heavily refactored the TileLayer code as it was difficult for me to follow.

### CODE REVIEW

File: lib/src/layer/tile_layer/tile_manager.dart [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  void abortLoading(double? tileZoom, EvictErrorTileStrategy evictionStrategy) {
+    final toRemove = <String>[];
+    for (final entry in _tiles.entries) {
+      final tile = entry.value;
+
+      if (tile.coords.z != tileZoom) {
+        if (tile.loaded == null) {
+          toRemove.add(entry.key);
+        }
+      }
```
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_manager.dart:18]: Using nested if statements is generally worse for performance than a logic gate, especially where readability will not be affected that much.

```suggestion
      if (tile.coords.z != tileZoom && tile.loaded == null) {
        toRemove.add(entry.key);
      }
```

File: lib/src/layer/tile_layer/tile_layer_options.dart [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+  /// Tile image to show in place of the tile that failed to load.
+  final ImageProvider? errorImage;
+
+  /// Static information that should replace placeholders in the [urlTemplate].
+  /// Applying API keys is a good example on how to use this parameter.
+  ///
+  /// Example:
+  ///
+  /// ```dart
+  ///
```
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_layer_options.dart:139]: Not sure if this was already here, but this empty line can be removed.

```suggestion
```

File: lib/src/layer/tile_layer/tile_layer_options.dart [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+  /// TileLayerOptions(
+  ///     urlTemplate: "https://api.tiles.mapbox.com/v4/"
+  ///                  "{id}/{z}/{x}/{y}{r}.png?access_token={accessToken}",
+  ///     additionalOptions: {
+  ///         'accessToken': '<PUT_ACCESS_TOKEN_HERE>',
+  ///          'id': 'mapbox.streets',
+  ///     },
+  /// ),
+  /// ```
+  ///
```
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_layer_options.dart:149]: Not sure if this was already here, but this empty line can be removed.

```suggestion
```

File: lib/src/layer/tile_layer/tile_layer.dart [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+  TileLayer({
+    required this.options,
+    required this.mapState,
+    required this.stream,
+  }) : super(key: options.key);
+
+  @override
+  State<StatefulWidget> createState() {
+    return _TileLayerState();
+  }
```
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_layer.dart:53]: Removing unnecessary statements.

```suggestion
  State<StatefulWidget> createState() => _TileLayerState();
```

File: lib/src/layer/tile_layer/tile_layer.dart [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  CustomPoint getTileSize() {
+    return _tileSize;
+  }
```
Comment (JaffaKetchup) [inline lib/src/layer/tile_layer/tile_layer.dart:230]: Removing unnecessary statements.

```suggestion
  CustomPoint getTileSize() => _tileSize;
```

**Review by JaffaKetchup — CHANGES_REQUESTED** [review]
Haven't tested/run anything, but this is all I could see from scrolling through the changes! You've really done a great job, thanks :)
All the suggestions are just slight optimisations and tidy up, and I'm open to discussion if you disagree with any of them.

**Comment by rorystephenson** [comment]
Thanks @JaffaKetchup I've made those changes 👌 !

**Comment by JaffaKetchup** [comment]
Thanks. I'll do a full test at some point today!

**Review by ibrierley — APPROVED** [review]
All looks good to me. Great stuff, and thanks for all the work done on this.

**Comment by JaffaKetchup** [comment]
Testing now...

**Review by JaffaKetchup — APPROVED** [review]
LGTM. Many thanks once again!

**Comment by rorystephenson** [comment]
Awesome I wasn't expecting it to get reviewed and merged so quick, thanks to both of you!

**Pull request merged**